src/object/C4Object.cpp

```diff
@@ -280,12 +280,19 @@ void C4Object::AssignRemoval(bool exit_contents)
 
 	// ----------------------------------------------------------
 	// Remove particles, no destruction check necessary
-	ClearParticleLists();
+	ClearParticleLists(); // TODO: Move this to FinishRemoval(), too?
```

Comment:
That kind of depends on what the functions are supposed to do, which isn't clear from the names (AssignRemoval vs. FinishRemoval) - and they are also not commented. So for an outsider it is very unclear how to answer this question

Reply:
Yes, the TODO was more or less for the reviewer. More on that in the other comment, where FinishRemoval() was introduced.

Reply:
Even the reviewer needs to know what the function is supposed to do - and that should be reflected in the source. Either with a descriptive function name or with a comment. Because if it's ambiguous where something should go, you can be sure that someone puts it into the wrong place (e.g. the next time someone needs a place for a cleanup method and thinks about whether to put it into FinishRemoval or AssignRemoval).
Maybe if the functions have a different premise (e.g. a certain object status) that can be violated if people use it incorrectly, this premise could be reflected in some assertions?

Reply:
Yes, the comments were added to the header file in the latest commit. From my point of view, since clearing the particles does not rely on callbacks and does not change the object state, it makes more sense to do it later when the object is actually removed. 

The more pressing issue here is, that the PR implementation breaks scenario sections, see the comment after my additional commits. I am unsure how to address this.

At first, the fix seemed pretty simple, but it is not :)